### ISSUE TITLE
Update requirejs for a speed boost in builds

### ISSUE THREAD

**Issue by samreid** [phetsims/chipper#532]
I noticed this RequireJS issue which reported an optimization for builds https://github.com/requirejs/r.js/pull/900

I tested performance of grunt with requirejs ~2.1.8 compared to ~2.3.2, and saw these times:

2.1.8
real	0m49.580s
real	0m47.528s
real	0m47.381s

average:  48.163

2.3.2
real	0m34.031s
real	0m34.007s
real	0m32.701s

average: 33.57966

On average, the new version is 14.5 seconds faster per build, which is 30% of the entire build time.

I spot tested the built Function Builder on Chrome and did not spot any problems.

This seems like worth investigating because it will speed up each of our builds (including during build tests and production deploys).

**Comment by samreid** [comment]
I'd like to commit this change to master, but there are some considerations:

1. Other developers won't be using this until they npm update in chipper
2. The build-server will automatically start using this new version
3. The new version has not been thoroughly tested.

It would be nice if developers can start using this new version, to start looking for issues and to start benefiting from 30% faster build times.  The main risk I see is if there is an upcoming RC or Dev deploy which needs to be done quickly without incurring risks from a new requirejs version.  Sims that already have RC branches + SHAs will not pick up this change.

Assigning to @ariel-phet and @pixelzoom for discussion.

**Comment by ariel-phet** [comment]
@samreid I think it would be best to wait on this improvement until we have our next dev meeting so everyone is on the same page. We do have a couple of potential RC's upcoming that would be best to perform without incurring this risk (Capacitor Lab: Basics and States of Matter:Basics...SOM:B definitely)

**Comment by pixelzoom** [comment]
Deferring to @ariel-phet's decision, but I do have some questions.  @samreid said:

> I'd like to commit this change to master, 

It's not clear what "this change" is. Is this just an update to a newer version of requirejs? Are there any other changes required to PhET code or tool chain?

**Comment by samreid** [comment]
The change to commit would be in chipper/package.json:

```js
"requirejs": "~2.1.8",
```

would become

```
"requirejs": "~2.3.2",
```

**Comment by pixelzoom** [comment]
OK, so this is just a matter of updating to a newer (latest?) version of requirejs.  It seems like that could be tested by running the complete test-server suite.

**Comment by samreid** [comment]
Here are the results from the test-server:

![image](https://cloud.githubusercontent.com/assets/679486/21446556/d1a535be-c884-11e6-8d63-944f16a4943e.png)
![image](https://cloud.githubusercontent.com/assets/679486/21446560/de59572c-c884-11e6-82d6-031b84356055.png)
![image](https://cloud.githubusercontent.com/assets/679486/21446562/e2e6601e-c884-11e6-8694-61b31a41f452.png)

Sim errors (dev):

gravity-force-lab-basics
Uncaught TypeError: Cannot read property 'createTandem' of undefined
TypeError: Cannot read property 'createTandem' of undefined
    at GravityForceLabBasicsModel.GravityForceLabModel [as constructor] (http://localhost/gravity-force-lab/js/gravity-force-lab/model/GravityForceLabModel.js?bust=1482461713681:41:21)
    at new GravityForceLabBasicsModel (http://localhost/gravity-force-lab-basics/js/gravity-force-lab-basics/model/GravityForceLabBasicsModel.js?bust=1482461713681:19:26)
    at Screen.createModel (http://localhost/gravity-force-lab-basics/js/gravity-force-lab-basics-main.js?bust=1482461713681:46:20)
    at Screen.initializeModel (http://localhost/joist/js/Screen.js?bust=1482461713681:186:26)
    at Array. (http://localhost/joist/js/Sim.js?bust=1482461713681:588:18)
    at http://localhost/joist/js/Sim.js?bust=1482461713681:599:27
Grunt errors:

gravity-force-lab-basics
Running "eslint:allFiles" (eslint) task

Running "clean" task

Running "requirejs-build" task
requirejs failed
Error: Error: Loader plugin did not call the load callback in the build:
image:
  image!GRAVITY_FORCE_LAB_BASICS/background.png: Error: unacceptable license entry for GRAVITY_FORCE_LAB_BASICS/background.png
Module loading did not complete for: gravity-force-lab-basics-main, GRAVITY_FORCE_LAB_BASICS/gravity-force-lab-basics/view/GravityForceLabBasicsScreenView

Running "after-requirejs-build" task
>> Unused images file: GRAVITY_FORCE_LAB_BASICS/background.png
Fatal error: gravity-force-lab is missing from phetLibs in package.json

**Comment by pixelzoom** [comment]
Hmmm... There's an error message for gravity-force-lab-basics, which has a red box ("failure after loading"). But there are also multiple runnables with an orange box ("failure loading, before running sim").  Were there no associated error messages for these?

**Comment by samreid** [comment]
> Were there no associated error messages for these?

No.  And I've seen this issue before where there is a problem with no textual report, see https://github.com/phetsims/aqua/issues/14

**Comment by jonathanolson** [comment]
> The build-server will automatically start using this new version

The build server should check out SHAs that would reference the old version, correct? So it wouldn't do this except for new RCs that explicitly use it?

Also, does this update uglify (maybe that's why there's a speed increase), and/or can we update uglify? I recall uglification took up most of the build time fairly recently.

**Comment by samreid** [comment]
Right, the build server should check out SHAs then run `npm update`, so it would only use the new version for new publications.

The speed increase is documented here: https://github.com/requirejs/r.js/pull/900 It says "Improve traverse and parse performance" for the r.js step.

**Comment by jonathanolson** [comment]
If we're updating require.js, can we also look at updating uglify?

**Comment by samreid** [comment]
@samreid will update the requirejs on his working copy and re-run tests.  If tests for new requirejs are better than or equal tests for old requirejs, I'll upgrade.

**Comment by samreid** [comment]
I'll try out "uglify-js": "~2.7.5" while I'm at it.

**Comment by samreid** [comment]
I'm having the same problem where the test reports as failed, but there is no corresponding text (and no clues in the console):

![image](https://cloud.githubusercontent.com/assets/679486/21698101/bf2ee672-d352-11e6-8645-c354ac1b763a.png)

**Comment by samreid** [comment]
Here's the latest full run.  I have no idea why the rightmost 2 columns are blank.

![image](https://cloud.githubusercontent.com/assets/679486/21699044/22faf6ba-d357-11e6-8a98-991eb126f056.png)
![image](https://cloud.githubusercontent.com/assets/679486/21699053/2b0be616-d357-11e6-927a-117828fbdbc8.png)
![image](https://cloud.githubusercontent.com/assets/679486/21699060/323f9f7c-d357-11e6-91da-e2cacc1133d8.png)

**Comment by samreid** [comment]
@jonathanolson also recommends to check speeds before and after upgrading uglify.

**Comment by samreid** [comment]
I'm having a lot less trouble with `testDuration=20000&testConcurrentBuilds=1`,  thanks @jonathanolson!

**Comment by samreid** [comment]
Tests with 
```js
    "uglify-js": "~2.7.5",
    "requirejs": "~2.3.2",
```

![image](https://cloud.githubusercontent.com/assets/679486/21701558/784c51b6-d363-11e6-81e6-070155033d4f.png)
![image](https://cloud.githubusercontent.com/assets/679486/21701562/7fc5ad5c-d363-11e6-88f9-ef956fd986e6.png)
![image](https://cloud.githubusercontent.com/assets/679486/21701592/ac6c0acc-d363-11e6-95b9-fa4a75d5a3b2.png)

```
Grunt errors:

gravity-force-lab-basics
Running "eslint:allFiles" (eslint) task

Running "clean" task

Running "requirejs-build" task
requirejs failed
Error: Error: Loader plugin did not call the load callback in the build:
image:
  image!GRAVITY_FORCE_LAB_BASICS/background.png: Error: unacceptable license entry for GRAVITY_FORCE_LAB_BASICS/background.png
Module loading did not complete for: gravity-force-lab-basics-main, GRAVITY_FORCE_LAB_BASICS/gravity-force-lab-basics/view/GravityForceLabBasicsScreenView

Running "after-requirejs-build" task
>> Unused images file: GRAVITY_FORCE_LAB_BASICS/background.png
Fatal error: gravity-force-lab is missing from phetLibs in package.json
```

This looks like a GFLB problem, not a problem with requirejs or uglify.

**Comment by samreid** [comment]
I've updated requirejs and uglify in the preceding commit.  Developers will need to pull chipper then `npm update` to get the changes.

@jonathanolson suggested to review performance changes in uglify2.  I'll create a separate issue for that.

**Comment by jonathanolson** [comment]
Presumably all active-runnable/scenery/kite/dot package.json files would also need to be updated?

**Comment by jonathanolson** [comment]
For example, acid-base-solutions still seems to be on:
```js
  "devDependencies": {
    "grunt": "~1.0.0",
    "grunt-eslint": "~19.0.0",
    "uglify-js": "~2.4.0",
    "requirejs": "~2.1.8"
  },
```

**Comment by samreid** [comment]
@jonathanolson I think simulations are using chipper requirejs.  I tried the following steps:

1. Remove '    "requirejs": "~2.1.8"' from acid-base-solutions/package.json
2. cd acid-base-solutions
3. rm -rf node_modules
4. npm install
5. grunt

And it succeeds.  However, if I try the same thing with uglify, there is a failure with uglify not found:

```
~/github/acid-base-solutions$ grunt
Running "eslint:allFiles" (eslint) task

Running "clean" task

Running "requirejs-build" task

Running "after-requirejs-build" task
Fatal error: Cannot find module '../../../acid-base-solutions/node_modules/uglify-js'
```

Perhaps sim builds use chipper requirejs but simulation uglify.  If we can point to the chipper uglify, it will take us another step toward our dream of smaller node_modules.

**Comment by jonathanolson** [comment]
> Perhaps sim builds use chipper requirejs but simulation uglify.

Probably due to the following:
```js
  // TODO: chipper#101 eek, this is scary! we are importing from the repository dir. ideally we should just have uglify-js installed once in chipper?
  var uglify = require( '../../../' + buildConfig.name + '/node_modules/uglify-js' );
```

We could spend a bit more time trying to normalize this (https://github.com/phetsims/chipper/issues/101), or even replace it with:
```js
var uglify = require( '../../../chipper/node_modules/uglify-js' );
```

> If we can point to the chipper uglify, it will take us another step toward our dream of smaller node_modules.

That sounds great!

**Comment by jonathanolson** [comment]
> I think simulations are using chipper requirejs.

Can we move forward with removing this from sims? Might require an independent check for scenery/kite/dot too.

**Comment by samreid** [comment]
I'll work on removing sim requirejs and uglify in a separate issue.  But this revelation invalidates the results in https://github.com/phetsims/chipper/issues/532#issuecomment-270788287, if they were indeed using the sim's uglify.

**Comment by samreid** [comment]
I switched to use chipper's uglify in https://github.com/phetsims/chipper/issues/538.

**Comment by samreid** [comment]
Running a 5000ms test, there are no systemic issues:

![image](https://cloud.githubusercontent.com/assets/679486/21946833/6502c468-d99f-11e6-94b3-74fe020bae19.png)

Uglify2 performance will be tested in https://github.com/phetsims/chipper/issues/536, I think this issue is ready to be closed.
